File: packages/python/src/quanttide_course/models/lecture.py
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, field_validator


class Level(str, Enum):
    introductory = "初级"
    intermediate = "中级"
    advanced = "高级"


class Lecture(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    id: str
    title: str
    description: str
    level: Level
    targets: list[str]
    objectives: list[str]
    points: list[str]
    level: Level

    @field_validator("targets", "objectives", "points", mode="before")
    @classmethod
    def coerce_list(cls, v: Any) -> list[str]:
        if v is None:
            return []
        if isinstance(v, list):
            return [str(item) for item in v]
        return []

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump(mode="json")

File: packages/dart/lib/src/lecture.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'lecture.freezed.dart';
part 'lecture.g.dart';

enum Level {
  @JsonValue('初级') introductory,
  @JsonValue('中级') intermediate,
  @JsonValue('高级') advanced;
}

@freezed
class Lecture with _$Lecture {
  const factory Lecture({
    required String id,
    required String title,
    required String description,
    required List<String> targets,
    required List<String> objectives,
    required List<String> points,
    required Level level,
  }) = _Lecture;

  factory Lecture.fromJson(Map<String, dynamic> json) => _$LectureFromJson(json);
}
